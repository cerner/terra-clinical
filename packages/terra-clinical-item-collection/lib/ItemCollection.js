'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _resizeObserverPolyfill = require('resize-observer-polyfill');

var _resizeObserverPolyfill2 = _interopRequireDefault(_resizeObserverPolyfill);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _breakpoints = require('terra-responsive-element/lib/breakpoints');

var _breakpoints2 = _interopRequireDefault(_breakpoints);

require('terra-base/lib/baseStyles');

require('./ItemCollection.scss');

var _CreateListView = require('./CreateListView');

var _CreateListView2 = _interopRequireDefault(_CreateListView);

var _CreateTableView = require('./CreateTableView');

var _CreateTableView2 = _interopRequireDefault(_CreateTableView);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var propTypes = {
  /**
  * The breakpoint to switch from a table view to a list view. Breakpoint options are tiny, small, medium, large, or huge.
  */
  breakpoint: _propTypes2.default.oneOf(['tiny', 'small', 'medium', 'large', 'huge']),
  /**
   * The styles to spread to the table. Table style options are isPadded and isStriped.
   **/
  tableStyles: _propTypes2.default.shape({
    isPadded: _propTypes2.default.bool,
    isStriped: _propTypes2.default.bool
  }),
  /**
   * The styles to spread to the list. List style options are isDivided and hasChevrons.
   **/
  listStyles: _propTypes2.default.shape({
    isDivided: _propTypes2.default.bool,
    hasChevrons: _propTypes2.default.bool
  }),
  /**
   * The array of hashes to be displayed as rows. Each hash can contain a startAccessory, endAccessory,
   * comment, array of displays and a itemStyles hash. The item style options are layout, textEmphasis,
   * isTruncated and accessoryAlignment.
   **/
  rows: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    startAccessory: _propTypes2.default.element,
    displays: _propTypes2.default.arrayOf(_propTypes2.default.element),
    comment: _propTypes2.default.element,
    endAccessory: _propTypes2.default.element,
    itemStyles: _propTypes2.default.shape({
      layout: _propTypes2.default.oneOf(['oneColumn', 'twoColumns']),
      textEmphasis: _propTypes2.default.oneOf(['default', 'start']),
      isTruncated: _propTypes2.default.bool,
      accessoryAlignment: _propTypes2.default.oneOf(['alignTop', 'alignCenter'])
    })
  })),
  /**
   * A callback event that will be triggered when selection state changes.
   */
  onChange: _propTypes2.default.func
};

var defaultProps = {
  listStyles: undefined,
  tableStyles: undefined,
  rows: [],
  breakpoint: 'small',
  onChange: undefined
};

var ItemCollection = function (_React$Component) {
  _inherits(ItemCollection, _React$Component);

  function ItemCollection(props) {
    _classCallCheck(this, ItemCollection);

    var _this = _possibleConstructorReturn(this, (ItemCollection.__proto__ || Object.getPrototypeOf(ItemCollection)).call(this, props));

    _this.state = { display: 'table', selectedIndex: -1 };
    _this.setContainer = _this.setContainer.bind(_this);
    _this.handleResize = _this.handleResize.bind(_this);
    _this.handleSelection = _this.handleSelection.bind(_this);
    return _this;
  }

  _createClass(ItemCollection, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      this.resizeObserver = new _resizeObserverPolyfill2.default(function (entries) {
        _this2.handleResize(entries[0].contentRect.width);
      });
      this.resizeObserver.observe(this.container);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.resizeObserver.disconnect(this.container);
      this.container = null;
    }
  }, {
    key: 'setContainer',
    value: function setContainer(node) {
      if (node === null) {
        return;
      } // Ref callbacks happen on mount and unmount, element will be null on unmount
      this.container = node.parentNode;
    }
  }, {
    key: 'handleSelection',
    value: function handleSelection(event, selectedIndex) {
      this.setState({ selectedIndex: selectedIndex });
      if (this.props.onChange) {
        this.props.onChange(event, selectedIndex);
      }
    }
  }, {
    key: 'handleResize',
    value: function handleResize(width) {
      var display = void 0;
      var breakpoints = (0, _breakpoints2.default)();
      if (width < breakpoints[this.props.breakpoint]) {
        display = 'list';
      } else {
        display = 'table';
      }

      if (this.state.display !== display) {
        this.setState({ display: display });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          listStyles = _props.listStyles,
          tableStyles = _props.tableStyles,
          rows = _props.rows,
          breakpoint = _props.breakpoint,
          customProps = _objectWithoutProperties(_props, ['listStyles', 'tableStyles', 'rows', 'breakpoint']);

      if (!rows || !rows.length) {
        return null;
      }

      var attributes = _extends({}, customProps);
      attributes.className = (0, _classnames2.default)(['terraClinical-ItemCollection', 'terraClinical-ItemCollection--' + this.state.display + 'View', attributes.className]);

      var collectionDisplay = void 0;
      if (this.state.display === 'table') {
        collectionDisplay = (0, _CreateTableView2.default)(rows, tableStyles, this.state.selectedIndex, this.handleSelection);
      } else if (this.state.display === 'list') {
        collectionDisplay = (0, _CreateListView2.default)(rows, listStyles, this.state.selectedIndex, this.handleSelection);
      }

      return _react2.default.createElement(
        'div',
        _extends({ ref: this.setContainer }, attributes),
        _react2.default.createElement(
          'h2',
          null,
          'Item Collection - ',
          this.state.display
        ),
        _react2.default.createElement(
          'h2',
          null,
          ' Item Selection is: ',
          this.state.selectedIndex,
          ' '
        ),
        collectionDisplay
      );
    }
  }]);

  return ItemCollection;
}(_react2.default.Component);

ItemCollection.propTypes = propTypes;
ItemCollection.defaultProps = defaultProps;

exports.default = ItemCollection;