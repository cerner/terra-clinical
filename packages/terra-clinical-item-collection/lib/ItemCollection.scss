@import '~terra-mixins';

.terraClinical-ItemCollection-column--accessory {
  width: 40px;
}
