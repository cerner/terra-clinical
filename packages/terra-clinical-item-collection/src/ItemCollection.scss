:local {
  .column-accessory,
  .list-accessory {
    width: 40px;
  }

  .content-accessory {
    vertical-align: middle !important;
  }
}
