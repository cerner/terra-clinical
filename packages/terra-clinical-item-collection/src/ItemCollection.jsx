import React from 'react';
import PropTypes from 'prop-types';
import ResponsiveElement from 'terra-responsive-element';
import 'terra-base/lib/baseStyles';
import './ItemCollection.scss';

import createListView from './_CreateListView';
import createTableView from './_CreateTableView';

const propTypes = {
  /**
  * The breakpoint to switch from a table view to a list view. Breakpoint options are tiny, small, medium, large, or huge.
  */
  breakpoint: PropTypes.oneOf(['tiny', 'small', 'medium', 'large', 'huge']),
  /**
   * The styles to spread to the table. Table style options are isPadded and isStriped.
   */
  tableStyles: PropTypes.shape({
    isPadded: PropTypes.bool,
    isStriped: PropTypes.bool,
  }),
  /**
   * The styles to spread to the list. List style options are isDivided and hasChevrons.
   */
  listStyles: PropTypes.shape({
    isDivided: PropTypes.bool,
    hasChevrons: PropTypes.bool,
  }),
  /**
   * The array of hashes to be displayed as rows. Each hash can contain a startAccessory, endAccessory,
   * comment, array of displays, isSelected, and a itemStyles hash. The item style options are layout,
   * textEmphasis, isTruncated and accessoryAlignment.
   */
  rows: PropTypes.arrayOf(PropTypes.shape({
    startAccessory: PropTypes.element,
    displays: PropTypes.arrayOf(PropTypes.element),
    comment: PropTypes.element,
    endAccessory: PropTypes.element,
    isSelected: PropTypes.bool,
    itemStyles: PropTypes.shape({
      layout: PropTypes.oneOf(['oneColumn', 'twoColumns']),
      textEmphasis: PropTypes.oneOf(['default', 'start']),
      isTruncated: PropTypes.bool,
      accessoryAlignment: PropTypes.oneOf(['alignTop', 'alignCenter']),
    }),
  })),
  /**
   * A callback event that will be triggered when selection state changes.
   */
  onChange: PropTypes.func,
};

const defaultProps = {
  listStyles: undefined,
  tableStyles: undefined,
  rows: [],
  breakpoint: 'small',
  onChange: undefined,
};

class ItemCollection extends React.Component {
  static selectedIndexFromRows(rows) {
    if (!rows || !rows.length) {
      return -1;
    }
    for (let i = 0; i < rows.length; i += 1) {
      if (rows[i].isSelected) {
        return i;
      }
    }
    return -1;
  }

  constructor(props) {
    super(props);
    this.state = { display: 'table', selectedIndex: ItemCollection.selectedIndexFromRows(this.props.rows) };
    this.handleSelection = this.handleSelection.bind(this);
  }

  handleSelection(event, selectedIndex) {
    this.setState({ selectedIndex });
    if (this.props.onChange) {
      this.props.onChange(event, selectedIndex);
    }
  }

  render() {
    const { listStyles, tableStyles, rows, breakpoint, ...customProps } = this.props;

    if (!rows || !rows.length) {
      return null;
    }

    const tableView = createTableView(rows, tableStyles, this.state.selectedIndex, this.handleSelection);
    const tableDisplay = (
      <div {...customProps} data-terra-clinical-item-collection-table-view className={customProps.className}>
        {tableView}
      </div>
    );

    const listView = createListView(rows, listStyles, this.state.selectedIndex, this.handleSelection);
    const listDisplay = (
      <div {...customProps} data-terra-clinical-item-collection-list-view className={customProps.className}>
        {listView}
      </div>
    );

    const breakpointDisplay = {};
    breakpointDisplay[breakpoint] = tableDisplay;

    return (
      <ResponsiveElement
        responsiveTo="parent"
        defaultElement={listDisplay}
        {...breakpointDisplay}
      />
    );
  }
}

ItemCollection.propTypes = propTypes;
ItemCollection.defaultProps = defaultProps;

export default ItemCollection;
