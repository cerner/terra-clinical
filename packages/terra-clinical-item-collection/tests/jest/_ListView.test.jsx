import React from 'react';
import ItemCollection from '../../src/ItemCollection';
import ListView from '../../src/_ListView';

const defaultRequiredElements = { hasStartAccessory: true, numberOfDisplays: 8, hasComment: true, hasEndAccessory: true };

it('should render a default ListView', () => {
  const item = (<ListView />);
  const component = shallow(item);
  expect(component).toMatchSnapshot();
});

it('should render a ListView with children', () => {
  const item = (
    <ListView requiredElements={defaultRequiredElements}>
      <ItemCollection.Item startAccessory={<p>S</p>} comment={<ItemCollection.Comment text="test comment" />} endAccessory={<p>E</p>}>
        <ItemCollection.Display text="Display 1" />
      </ItemCollection.Item>
    </ListView>
  );
  const component = shallow(item);
  expect(component).toMatchSnapshot();
});

it('should render a ListView with a divided list', () => {
  const item = (
    <ListView requiredElements={defaultRequiredElements} isListDivided>
      <ItemCollection.Item key="1" startAccessory={<p>S</p>} comment={<ItemCollection.Comment text="test comment" />} endAccessory={<p>E</p>}>
        <ItemCollection.Display text="Display 1" />
      </ItemCollection.Item>
    </ListView>
  );
  const component = shallow(item);
  expect(component).toMatchSnapshot();
});
