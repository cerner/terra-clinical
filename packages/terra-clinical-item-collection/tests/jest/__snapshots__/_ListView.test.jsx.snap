// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a ListView with a divided list 1`] = `
<List
  data-terra-clinical-item-collection-list-view={true}
  isDivided={true}
>
  <Item
    accessoryAlignment="alignCenter"
    comment={
      <ItemComment
        isTruncated={false}
        text="test comment"
      />
    }
    endAccessory={
      <p>
        E
      </p>
    }
    isListItemTruncated={false}
    isSelectable={false}
    isSelected={false}
    key=".$1"
    listItemLayout="oneColumn"
    listItemTextEmphasis="default"
    reserveStartAccessorySpace={true}
    showListItemChevron={false}
    startAccessory={
      <p>
        S
      </p>
    }
    view="list"
  >
    <ItemDisplay
      iconAlignment="center"
      isDisabled={false}
      isTruncated={false}
      text="Display 1"
    />
  </Item>
</List>
`;

exports[`should render a ListView with children 1`] = `
<List
  data-terra-clinical-item-collection-list-view={true}
  isDivided={false}
>
  <Item
    accessoryAlignment="alignCenter"
    comment={
      <ItemComment
        isTruncated={false}
        text="test comment"
      />
    }
    endAccessory={
      <p>
        E
      </p>
    }
    isListItemTruncated={false}
    isSelectable={false}
    isSelected={false}
    key=".0"
    listItemLayout="oneColumn"
    listItemTextEmphasis="default"
    reserveStartAccessorySpace={true}
    showListItemChevron={false}
    startAccessory={
      <p>
        S
      </p>
    }
    view="list"
  >
    <ItemDisplay
      iconAlignment="center"
      isDisabled={false}
      isTruncated={false}
      text="Display 1"
    />
  </Item>
</List>
`;

exports[`should render a default ListView 1`] = `
<List
  data-terra-clinical-item-collection-list-view={true}
  isDivided={false}
/>
`;
